Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/SurveyQuestions.kt

```diff
@@ -19,216 +19,74 @@ package com.example.compose.jetsurvey.survey
 import com.example.compose.jetsurvey.theme.stronglyDeemphasizedAlpha
-import com.google.accompanist.permissions.ExperimentalPermissionsApi
-import com.google.accompanist.permissions.MultiplePermissionsState
-import com.google.accompanist.permissions.rememberMultiplePermissionsState
-
-@OptIn(ExperimentalPermissionsApi::class)
-@Composable
-fun Question(
-    question: Question,
-    answer: Answer<*>?,
-    shouldAskPermissions: Boolean,
-    onAnswer: (Answer<*>) -> Unit,
-    onAction: (Int, SurveyActionType) -> Unit,
-    onDoNotAskForPermissions: () -> Unit,
-    modifier: Modifier = Modifier
-) {
-    if (question.permissionsRequired.isEmpty()) {
-        QuestionContent(question, answer, onAnswer, onAction, modifier)
-    } else {
-        val multiplePermissionsState = rememberMultiplePermissionsState(
-            question.permissionsRequired
-        )
-
-        if (multiplePermissionsState.allPermissionsGranted) {
-            QuestionContent(question, answer, onAnswer, onAction, modifier)
-        } else {
-            PermissionsRationale(
-                question,
-                multiplePermissionsState,
-                onDoNotAskForPermissions,
-                modifier.padding(horizontal = 20.dp)
-            )
-        }
-
-        // If we cannot ask for permissions, inform the caller that can move to the next question
-        if (!shouldAskPermissions) {
-            LaunchedEffect(true) {
-                onAnswer(Answer.PermissionsDenied)
-            }
-        }
-    }
-}
-
-@OptIn(ExperimentalPermissionsApi::class)
-@Composable
-private fun PermissionsRationale(
-    question: Question,
-    multiplePermissionsState: MultiplePermissionsState,
-    onDoNotAskForPermissions: () -> Unit,
-    modifier: Modifier = Modifier
-) {
-    Column(modifier) {
-        Spacer(modifier = Modifier.height(32.dp))
-        QuestionTitle(question.questionText)
-        Spacer(modifier = Modifier.height(32.dp))
-        val rationaleId =
-            question.permissionsRationaleText ?: R.string.permissions_rationale
-        Text(stringResource(id = rationaleId))
-        Spacer(modifier = Modifier.height(16.dp))
-        OutlinedButton(
-            onClick = {
-                multiplePermissionsState.launchMultiplePermissionRequest()
-            }
-        ) {
-            Text(stringResource(R.string.request_permissions))
-        }
-        Spacer(modifier = Modifier.height(8.dp))
-        OutlinedButton(onClick = onDoNotAskForPermissions) {
-            Text(stringResource(R.string.do_not_ask_permissions))
-        }
-    }
-}
 
+/**
+ * Creates a Column that is vertically scrollable and contains the title, the directions (if a
+ * string resource is passed), and the content
+ */
 @Composable
-private fun QuestionContent(
-    question: Question,
-    answer: Answer<*>?,
-    onAnswer: (Answer<*>) -> Unit,
-    onAction: (Int, SurveyActionType) -> Unit,
-    modifier: Modifier = Modifier
+fun QuestionWrapper(
```

Comment:
I'm wondering if this QuestionWrapper is really necessary, or if it's premature optimization and makes our survey less flexible. For example, what if I want to show a scrollable grid in one of the questions (or a lazy column for that matter), it would not work because there's already a scrollable column defined here.

What about making a reusable `QuestionTitle` and `QuestionDirections`. You can also save the modifier chains and reuse them, e.g. `questionTitleModifier`.

Reply:
That's actually how I originally did it (which is why QuestionTitle and QuestionDirections weren't private). Since Column is simple, I didn't mind having that portion repeated in each question, but then I realized I needed it to be a scrollable Column. Repeating that modifier (even if extracted) made it feel like I was requiring each question to build the UI with the same building blocks in the same way. Each time I was updating a question, I was basically just copying and pasting the same chunks from another question.

Maybe if we have cases where there's another layout like a grid, we'd rename this to something like ScrollableColumnQuestion and the new one might be VerticalGridQuestion or something like that? In that case, the QuestionTitle and QuestionDirections composables might need to be different anyway.

As a side note, the reason I went with a scrollable Column vs a LazyColumn was because the majority of each question is visible on most phones, so the lazy behavior wasn't a performance win. If we had a question with a dozen potential answers, then a LazyColumn would make more sense for that question, but then it gets back to whether this should be in a wrapper or in the individual questions or something else.